File: Java/Plasma/NucleusLib/pnSceneObject/plDrawInterface.java
package Plasma.NucleusLib.pnSceneObject;

import java.io.IOException;
import Plasma.CoreLib.hsStream;
import Plasma.CoreLib.hsTArray;
import Plasma.NucleusLib.pnKeyedObject.*;

public class plDrawInterface extends plObjInterface {
    public hsTArray<plDrawableKeyRef> Drawables = new hsTArray<plDrawableKeyRef>();
    public hsTArray<plKeyRef> Objects = new hsTArray<plKeyRef>();

    public void read(hsStream S) throws IOException {
        super.read(S);
        Drawables.read(S);
        Objects.read(S);
    }

    public void write(hsStream S) throws IOException {
        super.write(S);
        Drawables.write(S);
        Objects.write(S);
    }
}

File: C++/NucleusLib/pnSceneObject/plDrawInterface.cpp
#include "plDrawInterface.h"

/*plDrawableKeyRef */
plDrawableKey::plDrawableKey() : DrawKey(0) { }
plDrawableKey::~plDrawableKey() { }

void plDrawableKey::read(hsStream *S) {
    DrawKey = S->readInt();
    plKey::read(S);
}

void plDrawableKey::write(hsStream *S) {
    S->writeInt(DrawKey);
    plKey::write(S);
}

void plDrawableKey::readRef(hsStream *S) {
    S->writeInt(DrawKey);
    plKey::readRef(S);
}

void plDrawableKey::writeRef(hsStream *S) {
    S->writeInt(DrawKey);
    plKey::writeRef(S);
}


/* plDrawInterface */

plDrawInterface::plDrawInterface(PlasmaVer pv) { }
plDrawInterface::~plDrawInterface() { }

void plDrawInterface::read(hsStream *S) {
    plObjInterface::read(S);
    int count = S->readInt();
    Drawables.clear();
    Drawables.setSize(count);
    for (int i=0; i<count; i++)
        Drawables[i].readRef(S);
    count = S->readInt();
    Objects.clear();
    Objects.setSize(count);
    for (int i=0; i<count; i++)
        Objects[i].readRef(S);
}

void plDrawInterface::write(hsStream *S) {
    plObjInterface::write(S);
    S->writeInt(Drawables.getSize());
    for (int i=0; i<Drawables.getSize(); i++)
        Drawables[i].writeRef(S);
    S->writeInt(Objects.getSize());
    for (int i=0; i<Objects.getSize(); i++)
        Objects[i].writeRef(S);
}

